### DISCUSSION TITLE
To preview users: On upgrading to stable 9.0.0

### DISCUSSION THREAD

**Discussion by DuCanhGH** [serwist/serwist#127] [Announcements]
Serwist 9.0.0 has been released! Here is the release note: https://serwist.pages.dev/blog/2024/03/10/serwist-v9.

Throughout its development, there had been various experiments on how the API would look like in the end. As such, the preview builds had breaking changes between one another. This discussion lists all of such changes, aiming to help you migrate from one of them to the stable build.

- Preview 0 -> Preview 1:

  - `@serwist/next`: Renamed `cacheOnFrontEndNav` to `cacheOnNavigation`
    - See https://serwist.pages.dev/blog/2024/03/10/serwist-v9#renamed-cofen-to-con for more details.

- Preview 2 -> Preview 3:

  - `@serwist/sw`: Removed support for string handlers in `registerRuntimeCaching`
    - See https://serwist.pages.dev/blog/2024/03/10/serwist-v9#removed-string-handlers for more details.

- Preview 3 -> Preview 4:

  - `@serwist/precaching`: Renamed `fallbackURL`, added support for a `matcher`

- Preview 6 -> Preview 7:

  - `@serwist/svelte`: Moved Svelte integration into a separate package
    - See https://serwist.pages.dev/blog/2024/03/10/serwist-v9#moved-svelte-integration for more details.

- Preview 7 -> Preview 8:

  - `@serwist/core`: Allow non-Promise return types for `SerwistPlugin` callbacks

- Preview 12 -> Preview 13:

  - `@serwist/build`: Moved framework-specific types out of `@serwist/build`
    - See https://serwist.pages.dev/blog/2024/03/10/serwist-v9#moved-framework-types for more details.

- Preview 13 -> Preview 14:

  - `@serwist/cli`: Removed/renamed certain features
    - `copyLibraries` was already a no-op, so this simply removes the empty command.
    - `wizard --injectManifest` has become `wizard`, thanks to the fact that GenerateSW no longer exists.
    - Renamed the `injectManifest` command to `inject-manifest`. From now on, the CLI uses the `kebab-case` naming convention rather than the `camelCase` one.

- Preview 16 -> Preview 17:

  - `@serwist/sw`: Added `Serwist`
    - This class was initially meant to only replace `installSerwist`. Back then, its constructor only accepted two arguments, which are `precacheController` and `router`. Its then `install` function was the equivalent of `installSerwist`.

- Preview 17 -> Preview 18:
  - `@serwist/sw`: Merged service worker packages into `@serwist/sw`:
    - These service worker modules were merged into `@serwist/sw`:
      - Modules then located at `@serwist/sw/plugins`:
        - `@serwist/background-sync`
        - `@serwist/broadcast-update`
        - `@serwist/cacheable-response`
        - `@serwist/expiration`
        - `@serwist/google-analytics`
        - `@serwist/range-requests`
      - Modules then located at `@serwist/sw/precaching`:
        - `@serwist/precaching`:
          - This excludes `PrecacheFallbackPlugin`, `PrecacheFallbackEntry`, and `PrecacheFallbackPluginOptions`, which were located at `@serwist/sw/plugins`.
      - Modules then located at `@serwist/sw/routing`:
        - `@serwist/routing`
      - Modules then located at `@serwist/sw/strategies`:
        - `@serwist/strategies`
    - The legacy packages were and still are operable, but they are marked as deprecated.

- Preview 23 -> Preview 24:
  - `serwist`: Replaced `PrecacheController` and `Router` with `Serwist`
    - In this release, `@serwist/sw` was reverted to what it had been before Preview 17. Meeting the same fate as other service worker packages did, `@serwist/sw` has also been __deprecated__. `serwist` became what it was meant to become in 9.0.0.
    - `Serwist.install` was removed, and its functionality was merged with the class constructor's. At this point, rather than a `installSerwist` replacement, `Serwist` turned into the combination of `PrecacheController`, `Router`, and `installSerwist`. This became the final design of the class. See [the reference documentation on how to use it here](https://serwist.pages.dev/docs/serwist/core/serwist).
    - See [the discussion regarding this change](https://github.com/serwist/serwist/discussions/118) to learn more about the rationale behind it.
    - See https://serwist.pages.dev/blog/2024/03/10/serwist-v9#added-serwist for more details.

- Preview 25 -> Preview 26:
  - `serwist`: Internalized `cleanupOutdatedCaches` and `clientsClaim`.
    - There were little to no purpose in keeping these public, given that the `Serwist` class already provides the two options `precacheOptions.cleanupOutdatedCaches` and `clientsClaim`.
  - `@serwist/vite`: Moved `getSerwist` from `@serwist/vite/browser` to `virtual:serwist`
    - See https://serwist.pages.dev/blog/2024/03/10/serwist-v9#vite for more details.

This is literally me:

![image](https://github.com/serwist/serwist/assets/75556609/d3447ce2-bcf3-491e-8fb1-aa16393e2f1d)

Happy hacking!

P/s: happy 400!
![image](https://github.com/serwist/serwist/assets/75556609/61b8acfd-09b1-418a-a498-6216ad072ad4)

**Comment by maiconsanson** [comment]
Thank you for your big effort!
**The migration was successful.** 🥳 

I updated to:
```json
"@serwist/next": "9.0.0",
"@serwist/sw": "9.0.0",
```

But when I run `pnpm outdated`, I get the following:

<img width="427" alt="Screenshot 2024-04-22 at 11 20 43" src="https://github.com/serwist/serwist/assets/1965455/28ce099a-0a4b-4fcf-8122-59d8ad085a9e">

**Reply by DuCanhGH** [reply]
@maiconsanson yeah, you see, [the linked discussion in section "Preview 23 -> Preview 24"](https://github.com/serwist/serwist/discussions/118) states that:
 
> In v9.0.0, all of Serwist's service worker modules will be merged into one module named serwist. This module is now available on npm, and you can now install it and uninstall all the legacy modules: `npm uninstall @serwist/core @serwist/background-sync @serwist/broadcast-update @serwist/cacheable-response @serwist/core @serwist/expiration @serwist/google-analytics @serwist/navigation-preload @serwist/precaching @serwist/range-requests @serwist/routing @serwist/strategies @serwist/sw && npm install serwist`.
> 
> The legacy modules will be marked as deprecated on npm.

Basically, `serwist` has replaced all the service worker packages in v9.0.0. This is also mentioned in the blog: https://serwist.pages.dev/blog/2024/03/10/serwist-v9#merged-service-worker-packages. It will take just a little bit of effort to migrate, though. To migrate from `installSerwist`:

  - Old:

   ```ts
   import { installSerwist } from "@serwist/sw";
   import { defaultCache } from "@serwist/next/worker";
   
   installSerwist({
     precacheEntries: self.__SW_MANIFEST,
     runtimeCaching: defaultCache,
     // Other options...
   });
   ```

   - New:

   ```ts
   import { defaultCache } from "@serwist/next/worker";
   import { Serwist } from "serwist";
   
   const serwist = new Serwist({
     precacheEntries: self.__SW_MANIFEST,
     runtimeCaching: defaultCache,
     // Other options...
   });
   
   serwist.addEventListeners();
   ```

Or you may simply import `installSerwist` from `serwist/legacy`. Actually, the deprecated packages are now simply barrel files re-exporting modules from `serwist`. Not sure why you'd still use `installSerwist` when the migration takes just a few text replacements though. See the documentation here: https://serwist.pages.dev/docs/serwist/core/serwist.

Hope this helps!

**Reply by maiconsanson** [reply]
Yes, I already did this update too in the migration.
Oh my, just remove the `@serwist/sw` package.
That was dumb kkkk thank you!
